src/Orleans/Messaging/IncomingMessageBuffer.cs

```diff
@@ -0,0 +1,213 @@
+        private const int GROW_MAX_BLOCK_SIZE = 1024 * Kb; // 1mg
+        private readonly List<ArraySegment<byte>> readBuffer;
+        private readonly int maxSustainedBufferSize;
+        private int currentBufferSize;
+
+        private readonly byte[] lengthBuffer;
+
+        private int headerLength;
+        private int bodyLength;
+
+        private int recieveOffset;
+        private int decodeOffset;
+
+        private readonly bool supportForwarding;
+        private TraceLogger Log;
+
+        public IncomingMessageBuffer(TraceLogger logger, bool supportForwarding = false, int receiveBufferSize = DEFAULT_RECEIVE_BUFFER_SIZE, int maxSustainedReceiveBufferSize = DEFAULT_MAX_SUSTAINED_RECEIVE_BUFFER_SIZE)
+        {
+            Log = logger;
+            this.supportForwarding = supportForwarding;
+            currentBufferSize = receiveBufferSize;
+            maxSustainedBufferSize = maxSustainedReceiveBufferSize;
+            lengthBuffer = new byte[Message.LENGTH_HEADER_SIZE];
+            readBuffer = BufferPool.GlobalPool.GetMultiBuffer(currentBufferSize);
+            recieveOffset = 0;
+            decodeOffset = 0;
+            headerLength = 0;
+            bodyLength = 0;
+        }
+
+        public void UpdateReceivedData(int bytesRead)
+        {
+            recieveOffset += bytesRead;
+        }
+
+        public void Reset()
+        {
+            recieveOffset = 0;
+            decodeOffset = 0;
+            headerLength = 0;
+            bodyLength = 0;
+        }
+
+        public bool TryDecodeMessage(out Message msg)
+        {
+            msg = null;
+
+            // Is there enough read into the buffer to continue (at least read the lengths?)
+            if (recieveOffset - decodeOffset < KnownMessageSize())
+                return false;
+
+            // parse lengths if needed
+            if (headerLength == 0 || bodyLength == 0)
+            {
+                // get length segments
+                List<ArraySegment<byte>> lenghts = ByteArrayBuilder.BuildSegmentListWithLengthLimit(readBuffer, decodeOffset, Message.LENGTH_HEADER_SIZE);
+
+                // copy length segment to buffer
+                int lengthBufferoffset = 0;
+                foreach (ArraySegment<byte> seg in lenghts)
+                {
+                    Buffer.BlockCopy(seg.Array, seg.Offset, lengthBuffer, lengthBufferoffset, seg.Count);
+                    lengthBufferoffset += seg.Count;
+                }
+
+                // read lengths
+                headerLength = BitConverter.ToInt32(lengthBuffer, 0);
+                bodyLength = BitConverter.ToInt32(lengthBuffer, 4);
+            }
+
+            // If message is too big for default buffer size, grow
+            while (decodeOffset + KnownMessageSize() > currentBufferSize)
+            {
+                //TODO: Add configurable max message size for safety
```

Comment:
This is a good point on hardening the surface. I believe what ever is done should be documented elsewhere too, i.e. not only in code comment that is a bit hard to find.